SwiftyStoreKit/PaymentQueueController.swift

```diff
@@ -91,6 +94,26 @@ class PaymentQueueController: NSObject, SKPaymentTransactionObserver {
         paymentQueue.remove(self)
     }
 
+	var unhandledTransactions: [Purchase] {
+		guard let completeTransactions = completeTransactionsController.completeTransactions else {
+			print("SwiftyStoreKit.completeTransactions() should be called once when the app launches. Returning no results")
+			return []
+		}
+		let purchases = paymentQueue.transactions.flatMap { transaction -> Purchase? in
+			guard transaction.transactionState != .purchasing else {
+				return nil
+			}
+			return Purchase(
+				productId: transaction.payment.productIdentifier,
+				quantity: transaction.payment.quantity,
+				transaction: transaction,
+				originalTransaction: transaction.original,
+				needsFinishTransaction: !completeTransactions.atomically
```

Comment:
@huyduc Your use case seems a bit complex. I would suggest to do the following:

- Initiate the purchase with SwiftyStoreKit (you probably already do that)
- When you present the product page UI, use StoreKit directly to get the unfinished transactions as follows:
```swift
SKPaymentQueue.default().transactions
```
- Filter out any transactions in `.purchasing` state.
- Run your business logic as needed
- When you have delivered your purchase from the back-end, call either `SKPaymentQueue.default().finishTransaction(transaction)` or `SwiftyStoreKit.finishTransaction(transaction)` (they do exactly the same thing).

This solution bypasses SwiftyStoreKit entirely, but I think it makes sense for your special use case.

As a result of this, I'm not sure if it's possible for `finishTransaction` to be called twice (for example by SwiftyStoreKit.completeTransactions()`) and your explicit call. Maybe try to test this and see if everything work ok.

What do you think?